### PR TITLE
Disable python dispatcher in fallthrough for PyOperators

### PR DESCRIPTION
**Pull request by angelayi**
Possible fix for https://github.com/pytorch/pytorch/issues/89037

Context: The existing fallthrough implementation for PyOperators will cause the PythonDispatcher to infinitely redispatch to the PythonDispatcher due to this [line](https://github.com/pytorch/pytorch/blob/61fa43a1f2bd36b7f79333d83d989edb2a0353aa/c10/core/DispatchKeySet.h#L176) which permanently adds the PythonDispatcher to the dispatch key set which we get on this [line](https://github.com/pytorch/pytorch/blob/61fa43a1f2bd36b7f79333d83d989edb2a0353aa/torch/_ops.py#L195). We temporarily fixed this by excluding the PythonDispatcher key from the global keyset ([here](https://github.com/pytorch/pytorch/blob/61fa43a1f2bd36b7f79333d83d989edb2a0353aa/functorch/experimental/_cond.py#L156)), but this runs into an issue during the implementation for the functionalization key where we want to call [functionalize](https://github.com/pytorch/pytorch/blob/61fa43a1f2bd36b7f79333d83d989edb2a0353aa/functorch/experimental/_cond.py#L230-L231) for the true/false subgraphs, and [make_fx](https://github.com/pytorch/pytorch/blob/61fa43a1f2bd36b7f79333d83d989edb2a0353aa/functorch/experimental/_cond.py#L195) to check for aliasing/mutations, which requires having the PythonDispatcher key.

Our attempt at fixing this is to modify the fallthrough function to ignore the PythonDispatcher key when generating keys to redispatch to. This should prevent the infinite recursion, but won't modify the global state of having the PythonDispatcher key.

### CODE REVIEW

**Comment by ezyang** [comment]
how urgent is this

**Comment by angelayi** [comment]
> how urgent is this

semi-urgent? This is blocking turning on functionalization for DPE which uses the control flow ops. Right now we use a hacky version of functionalization which just skips the control flow ops, but it should be removed...

**Comment by ezyang** [comment]
I feel like there is probably a much simpler fix for the problem. Thinking.

File: functorch/experimental/_cond.py [hunk 1]
```diff
@@ -185,14 +179,15 @@ def _has_potential_branch_input_mutation(branch, fake_inputs):
...
+
+def _has_potential_branch_input_alias(branch, inputs):
     """
     Dispatch-trace the branch with fake inputs and check if
     producing graph has output aliasing the branch input. This is
     bit restrictive as the branch must be traceable.
     """
     try:
-        gm = make_fx(branch)(*fake_inputs)
+        gm = make_fx(branch)(*inputs)
```
Comment (ezyang) [inline functorch/experimental/_cond.py:190]: nit: try to avoid unrelated refactor like this, it makes it harder for reviewer to see what's going on

**Comment by ezyang** [comment]
see https://github.com/pytorch/pytorch/issues/89037#issuecomment-1453791918

File: functorch/experimental/_cond.py (hunk 1 above)
Comment (angelayi) [reply functorch/experimental/_cond.py:190]: Refactored those fixes into https://github.com/pytorch/pytorch/pull/95988, so those changes should disappear in this PR after that one merges!

**Comment by ezyang** [comment]
Hmmph. I don't like this, because you're smearing out state on the PyOperator that should be on a per call basis. And in fact, it's not even right, because if I fallthrough a key and then redispatch, I will *redo* that key (the fallthrough is not sticky!)

**Comment by ezyang** [comment]
Can you tell me more about why the approach we described in VC didn't work out?

**Comment by angelayi** [comment]
> Can you tell me more about why the approach we described in VC didn't work out?

My understanding from the VC is that order we want is for every key that is run (besides the PythonDispatcher), it should not be dispatched to again and go back to the PythonDispatcher to dispatch to the following key. So the order should look something like `PythonDispatcher -> PythonTLSSnapshot -> PythonDispatcher -> AutogradCPU -> PythonDispatcher ....`

The way I thought to do that was to use the ExcludeDispatchKeyGuard to prevent those keys from being dispatched to again. But because that affects the global set of keys, it prevented the inner make_fx call we make in cond from running correctly.

> if I fallthrough a key and then redispatch, I will redo that key (the fallthrough is not sticky!)

If you fallthrough a key wouldn't it get added to the list of keys that have been run already and redispatch to PythonDispatcher?

**Comment by ezyang** [comment]
> The way I thought to do that was to use the ExcludeDispatchKeyGuard to prevent those keys from being dispatched to again. But because that affects the global set of keys, it prevented the inner make_fx call we make in cond from running correctly.

Not necessary. Because python dispatcher can compute what the correct key to go to is. So then you just call it directly (op_dk, or just call the callable in your python side dispatch dict)

**Comment by ezyang** [comment]
No test?

**Comment by ezyang** [comment]
Please check whatever actual use case you needed isn't already fixed on master, I landed a set of orthogonal changes which should fix infinite fallthrough loops. I have no way of testing since this PR doesn't have a test.

**Comment by angelayi** [comment]
Yup, the actual case is fixed on master. https://github.com/pytorch/pytorch/pull/96635 to remove the existing hacky fallthrough (the tests were already landed previously).
